File: plugin-loader/src/main/java/com/tencent/cubershi/plugin_loader/classloaders/PluginClassLoader.java
package com.tencent.cubershi.plugin_loader.classloaders;

import dalvik.system.DexClassLoader;

/**
 * 用于加载插件的ClassLoader.
 *
 * @author cubershi
 */
public class PluginClassLoader extends DexClassLoader {


    private final ClassLoader mMockClassloader;
    private final String[] mMockClassNames;

    public PluginClassLoader(
            String dexPath, String optimizedDirectory, String librarySearchPath, ClassLoader parent,
            ClassLoader mockClassLoader,
            String[] mockClassNames
    ) {
        super(dexPath, optimizedDirectory, librarySearchPath, parent);
        mMockClassloader = mockClassLoader;
        mMockClassNames = mockClassNames;
    }

    @Override
    protected Class<?> loadClass(String className, boolean resolve) throws ClassNotFoundException {
        boolean isMockClass = false;
        for (String mockClassName : mMockClassNames) {
            if (className.equals(mockClassName)) {
                isMockClass = true;
                break;
            }
        }

        if (isMockClass) {
            return mMockClassloader.loadClass(className);
        } else {
            return super.loadClass(className, resolve);
        }
    }
}

File: plugin-loader/src/main/kotlin/com/tencent/cubershi/plugin_loader/classloaders/PluginClassLoader.kt
package com.tencent.cubershi.plugin_loader.classloaders

import android.content.Context
import com.tencent.hydevteam.pluginframework.pluginloader.multidex.MultiDexClassLoader


/**
 * 用于加载插件的ClassLoader.
 *
 * @author cubershi
 */
class PluginClassLoader(
        context: Context, dexPath: String, optimizedDirectory: String, private val librarySearchPath: String, parent: ClassLoader,
        private val mMockClassloader: ClassLoader,
        private val mMockClassNames: Array<String>
) : MultiDexClassLoader(context, dexPath, optimizedDirectory, librarySearchPath, parent) {

    fun getLibrarySearchPath() = librarySearchPath

    @Throws(ClassNotFoundException::class)
    override fun loadClass(className: String, resolve: Boolean): Class<*> {
        var isMockClass = false
        for (mockClassName in mMockClassNames) {
            if (className == mockClassName) {
                isMockClass = true
                break
            }
        }

        return if (isMockClass) {
            mMockClassloader.loadClass(className)
        } else {
            try {
                return super.loadClass(className, resolve)
            } catch (e: ClassNotFoundException) {
                //org.apache.commons.logging是非常特殊的的包,由系统放到App的PathClassLoader中.
                if (className.startsWith("org.apache.commons.logging")) {
                    return mMockClassloader.parent.loadClass(className)
                } else {
                    throw e
                }
            }
        }
    }
}
